asgard/projector.h

```diff
@@ -172,27 +177,70 @@ class Projector {
         if (projector_mode == "bss") {
             projector_mode = "walking";
         }
-        std::vector<valhalla::baldr::Location> locations;
-        std::transform(places_begin, places_end, std::back_inserter(locations),
-                       [this, &projector_mode](const valhalla::midgard::PointLL& place) {
-                           return build_location(place, min_outbound_reach, min_inbound_reach, radius, projector_mode);
-                       });
-        const auto path_locations = valhalla::loki::Search(locations,
-                                                           graph,
-                                                           costing);
 
-        std::unordered_map<valhalla::midgard::PointLL, valhalla::baldr::PathLocation> results;
-        for (const auto& l : path_locations) {
-            results.emplace(l.first.latlng_, l.second);
+        auto it_cache = cache_.find(projector_mode);
+        auto& matrix_cache = (it_cache != cache_.end()) ? it_cache->second : cache_["walking"];
+
+        const auto& matrix_map = matrix_cache.template get<1>();
+
+        auto& direct_path_list = direct_path_cache_.template get<0>();
+        auto& direct_path_map = direct_path_cache_.template get<1>();
+
+        {
+            std::lock_guard<std::mutex> lock(mutex);
+            for (auto it = places_begin; it != places_end; ++it) {
+                const auto search = matrix_map.find(std::make_pair(*it, projector_mode));
+                if (search != matrix_map.end()) {
+                    results.emplace(*it, search->second);
+                    continue;
+                }
+                missed_from_matrix.push_back(*it);
+            }
+        }
+
+        {
+            std::lock_guard<std::mutex> lock(direct_path_mutex);
+            for (const auto& l : missed_from_matrix) {
+                ++nb_dp_cache_calls_;
+                const auto direct_path_search = direct_path_map.find(std::make_pair(l, projector_mode));
+                if (direct_path_search != direct_path_map.end()) {
+                    direct_path_list.relocate(direct_path_list.begin(), direct_path_cache_.template project<0>(direct_path_search));
+                    results.emplace(l, direct_path_search->second);
+                    continue;
+                }
+                ++nb_dp_cache_miss_;
+                missed_from_dp.push_back(build_location(l, min_outbound_reach, min_inbound_reach, radius, projector_mode));
+            }
+        }
+
+        if (!missed_from_dp.empty()) {
+
+            const auto path_locations = valhalla::loki::Search(missed_from_dp,
+                                                               graph,
+                                                               costing);
+
+            std::lock_guard<std::mutex> lock(direct_path_mutex);
+            for (const auto& l : path_locations) {
+                results.emplace(l.first.latlng_, l.second);
+            }
+
```

Comment:
reduce lock duration
```suggestion
{
            std::lock_guard<std::mutex> lock(direct_path_mutex);
            for (const auto& l : path_locations) {
                results.emplace(l.first.latlng_, l.second);
            }
}
```